p4_symbolic/ir/ir.cc

```diff
@@ -0,0 +1,354 @@
+  for (const google::protobuf::Struct &primitive : bmv2_action.primitives()) {
+    if (primitive.fields().count("op") != 1 ||
+        primitive.fields().count("parameters") != 1) {
+      return absl::Status(
+          absl::StatusCode::kInvalidArgument,
+          absl::StrFormat("Primitive statement in action %s should contain 'op'"
+                          ", 'parameters', found %s",
+                          pdpi_action.preamble().name(),
+                          primitive.DebugString()));
+    }
+
+    const std::string &operation = primitive.fields().at("op").string_value();
+    // TODO(babman): Maybe bring back the enum and use switch-case here? discuss
+    Statement *statement = action_impl->add_action_body();
+    if (operation == "assign") {
+      AssignmentStatement *assignment = statement->mutable_assignment();
+      const google::protobuf::Value &params =
+          primitive.fields().at("parameters");
+      if (params.kind_case() != google::protobuf::Value::kListValue ||
+          params.list_value().values_size() != 2) {
+        return absl::Status(
+            absl::StatusCode::kInvalidArgument,
+            absl::StrFormat("Assignment statement in action %s must contain 2 "
+                            "parameters, found %s",
+                            pdpi_action.preamble().name(),
+                            primitive.DebugString()));
+      }
+
+      ASSIGN_OR_RETURN(
+          LValue lvalue,
+          ExtractLValue(params.list_value().values(0), variable_map));
+      ASSIGN_OR_RETURN(
+          RValue rvalue,
+          ExtractRValue(params.list_value().values(1), variable_map));
+      assignment->mutable_left()->CopyFrom(lvalue);
+      assignment->mutable_right()->CopyFrom(rvalue);
+    } else {
+      return absl::Status(
+          absl::StatusCode::kUnimplemented,
+          absl::StrFormat("Unsupported statement in action %s, found %s",
+                          pdpi_action.preamble().name(),
+                          primitive.DebugString()));
+    }
+
+    // Parse source_info struct into its own protobuf.
+    // Applies to all types of statements.
+    if (primitive.fields().count("source_info") != 1) {
+      return absl::Status(
+          absl::StatusCode::kInvalidArgument,
+          absl::StrFormat(
+              "Statement in action %s does not have source_info, found %s",
+              pdpi_action.preamble().name(), primitive.DebugString()));
+    }
+
+    ASSIGN_OR_RETURN(
+        bmv2::SourceLocation source_location,
+        ExtractSourceLocation(primitive.fields().at("source_info")));
+    statement->mutable_source_info()->CopyFrom(source_location);
+  }
+
+  return output;
+}
+
+// Parsing and validating tables.
+pdpi::StatusOr<Table> ExtractTable(
+    const bmv2::Table &bmv2_table,
+    const pdpi::ir::IrTableDefinition &pdpi_table) {
+  Table output;
+  // Table definition is copied from pdpi.
+  output.mutable_table_definition()->CopyFrom(pdpi_table);
```

Comment:
I can't access this link.
I am using CopyFrom in ceratin locations because when I tested the program with valgrind, there were some memory issues (double freeing) and copying fixed them. That does not apply to this particular instance, but it applies to some of the other ones.